src/frontend/src/helpers/Storage/WebStorage.ts

```diff
@@ -0,0 +1,588 @@
+    super();
+    this.connection = connection;
+    // this.synced = false;
+    this.connected = false;
+    this.failed = false;
+    // this.isSyncing = false;
+    this.offlineMode = 0; // TODO change this to look up the cookie
+    this.timeoutCount = 0;
+    this.timeoutInterval = null;
+    this.key = 0;
+    this.storage = storage;
+    this.socket = new SeashellWebsocket(this.connection.wsURI, this.connection.key);
+    this.callbacks = [];
+    this.debug = debug;
+    this.socket.debug = debug;
+
+    // These functions are not available in offline mode.
+    // this.ping = this.make_offline_disabled('ping');
+    // this.newProject = this.make_offline_disabled('newProject');/
+    // this.newProjectFrom = this.make_offline_disabled('newProjectFrom');
+    // this.deleteProject = this.make_offline_disabled('deleteProject');
+    // this.restoreFileFrom = this.make_offline_disabled('restoreFileFrom');
+    // this.getUploadFileToken = this.make_offline_disabled('getUploadFileToken');
+    // this.getExportToken = this.make_offline_disabled('getExportToken');
+    // this.marmosetSubmit = this.make_offline_disabled('marmosetSubmit');
+    // These functions do nothing and just resolve in offline mode.
+    // this.lockProject = this.make_offline_noop('lockProject');
+    // this.forceLockProject = this.make_offline_noop('forceLockProject');
+    // this.unlockProject = this.make_offline_noop('unlockProject');
+    // this.archiveProjects = this.make_offline_noop('archiveProjects');
+
+    // These functions either:
+    //  - return the online result if online.
+    //  - return the offline result if offline.
+    // this.getProjects = this.make_offline_enabled('getProjects');
+    // this.getFiles = this.make_offline_enabled('getFiles');
+    // this.readFile = this.make_offline_enabled('readFile');
+    // this.getRunFile = this.make_offline_enabled('getRunFile');
+    // this.getSettings = this.make_offline_enabled('getSettings');
+    // this.getMostRecentlyUsed = this.make_offline_enabled('getMostRecentlyUsed');
+
+    // These functions:
+    //  - invoke the offline version if offline
+    //  - invoke the both the offline version and the online
+    //    version if online, returning the offline version.
+    //  - marks the local storage as dirty and queues up a sync
+    // this.newFile = this.make_offline_enabled('newFile', true);
+    // this.writeFile = this.make_offline_enabled('writeFile', true);
+    // this.deleteFile = this.make_offline_enabled('deleteFile', true);
+    // this.renameFile = this.make_offline_enabled('renameFile', true);
+    // this.setSettings = this.make_offline_enabled('setSettings', true);
+    // this.updateMostRecentlyUsed = this.make_offline_enabled('updateMostRecentlyUsed', true)
+  }
+
+  public async getRunFile(proj: ProjectID, question: string): Promise<FileID> {
+    try {
+      const result = await this.socket.sendMessage({
```

Comment:
I'm actually thinking of splitting the run code compiler away from Storage. Like make it a separate class or something. And you can call `WebsocketClient.sendMessage` from that class for the compiling IO operation. I think `sendMessage` is high level enough to be used in other classes. Meanwhile I move the rest of interface functions from websocket to WebStorage. Although there won't be many, there will be some discrepancies between the current running websocket APIs and what I need here. For example, for listProject the current API returns a `[string,number,string][]`. I want to convert that to a `File[]`. It gets kind of ugly if I do it in `WebsocketClient`.

I also want to enforce the API interface with the backend. In that case I will make an interface.ts for it. I think WebsocketClient should just worry about handling authentication and sending message.